Userland/Libraries/LibGfx/JPEGLoader.cpp

```diff
@@ -437,9 +437,21 @@ static ErrorOr<void> build_macroblocks(JPEGLoadingContext& context, Vector<Macro
             for (u8 hfactor_i = 0; hfactor_i < scan_component.component.hsample_factor; hfactor_i++) {
                 // A.2.3 - Interleaved order
                 u32 mb_index = (vcursor + vfactor_i) * context.mblock_meta.hpadded_count + (hfactor_i + hcursor);
-                if (!context.current_scan.are_components_interleaved())
+                if (!context.current_scan.are_components_interleaved()) {
                     mb_index = vcursor * context.mblock_meta.hpadded_count + (hfactor_i + (hcursor * scan_component.component.vsample_factor) + (vfactor_i * scan_component.component.hsample_factor));
 
+                    // A.2.4 Completion of partial MCU
+                    // If the component is [and only if!] to be interleaved, the encoding process
+                    // shall also extend the number of samples by one or more additional blocks.
+
+                    // Horizontally
+                    if (mb_index >= context.mblock_meta.hcount && mb_index % context.mblock_meta.hpadded_count >= context.mblock_meta.hcount)
+                        continue;
+                    // Vertically
+                    if (mb_index >= context.mblock_meta.hpadded_count * context.mblock_meta.vcount)
```

Comment:
These comparisons (in both ifs) look fishy to me, but I might be thinking about them wrong. `mb_index` is a linearized index – shouldn't this compare based on vcursor and hcursor or something?

Also, why do we do a vertical check at all? The commit message only talks about horizontal alignment.

Reply:
They are fishy because, it's clearly a hack.

Some clarification: what I will call a big block is what we are iterating over when with the outer loop of [h, v] cursor.

In the classic scheme (interleaved components), and what was the decoder designed for, when reading the first big block, the data is written to macroblocks of indexes `0`, `1`, `line_length + 0`, `line_length + 1`. That is because components are interleaved.

However, in the other case data is ordered differently and corresponding macroblocks are simply `0`, `1`, `2`, `3`. This should be considered upstream, but this is not doable without refactoring a big part of the decoder. In 2c98eff, I added a way to convert the index back to a simple linear case. It exists a way to do this conversion perfectly (I mean, if there are bugs, it's only my fault; it's theoretically doable), still it has more of a hack. Indeed, in this context, `[h, v]_cursor` are meaningless, as you've seen even the first iteration starts to write to the second big block's land (2 and 3 are supposed to correspond to h_cursor = 2).

Now, to handle falsely-allocated macroblocks, we can't rely on cursors anymore, meaning that we are only left with mb_index to deduce the actual position of a block in the image. So yeah, it looks fishy to you, but also to everyone else, the whole `!are_components_interleaved` block is just a pile of hacks that accumulate until I'm confident enough with both our decoder and the spec to refactor this.

> Also, why do we do a vertical check at all? The commit message only talks about horizontal alignment.

Same reason as horizontal, the encoder will provide more blocks to complete columns if components are interleaved and the number of MCU per column is odd. I think that the commit message is not specific about horizontal alignment, as it talks about lines and columns. 

Anyway, let me know if something is still unclear (I tried to do my best :sweat_smile:) and if you think that I should modify the commit message.